### PR TITLE
add redis

### PR DESCRIPTION
**Pull request by BelehovEgor**

### CODE REVIEW

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs [hunk 1]
```diff
@@ -75,37 +84,30 @@ private List<DepartmentData> GetDepartments(List<DbProject> dbProjects, List<str
...
+      if (filter == null)
       {
-        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
-
-        return new FindResultResponse<ProjectInfo>
-        {
-          Status = OperationResultStatusType.Failed,
-          Errors = errors
-        };
+        throw new ArgumentNullException(nameof(filter));
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs:100]: should we throw?

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 2]
```diff
@@ -34,56 +35,46 @@ public class FindTasksCommand : IFindTasksCommand
...
-    private IGetUsersDataResponse GetUsersData(List<Guid> userId, List<string> errors)
+    private async Task<List<UserData>> GetUsersDatasThroughBroker(List<Guid> usersIds, List<string> errors)
     {
       string errorMessage = "Can not find user data. Please try again later.";
 
       try
       {
-        Response<IOperationResult<IGetUsersDataResponse>> response = _requestClient.GetResponse<IOperationResult<IGetUsersDataResponse>>(
-          IGetUsersDataRequest.CreateObj(userId)).Result;
+        var response = await _requestClient.GetResponse<IOperationResult<IGetUsersDataResponse>>(
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:66]: var

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 3]
```diff
@@ -113,37 +103,24 @@ private IGetUsersDataResponse GetUsersData(List<Guid> userId, List<string> error
...
-        && GetDepartment(userId, response.Errors)?.DirectorUserId != userId)
+      if (filter == null)
       {
-        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
-
-        response.Status = OperationResultStatusType.Failed;
-        response.Errors.Add("Not enough rights.");
-
-        return response;
+        throw new ArgumentNullException(nameof(filter));
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:113]: should we throw?

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 4]
```diff
@@ -113,37 +103,24 @@ private IGetUsersDataResponse GetUsersData(List<Guid> userId, List<string> error
...
       }
 
-      if (_findRequestValidator.ValidateCustom(filter, out List<string> errors))
-      {
-        _httpContextAccessor.HttpContext.Response.StatusCode = (int)HttpStatusCode.BadRequest;
+      List<string> errors = new();
 
-        response.Status = OperationResultStatusType.Failed;
-        response.Errors.AddRange(errors);
+      var userId = _httpContextAccessor.HttpContext.GetUserId();
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:118]: var

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 5]
```diff
@@ -113,37 +103,24 @@ private IGetUsersDataResponse GetUsersData(List<Guid> userId, List<string> error
...
+      var projectUsers = _userRepository.Find(userId).ToList();
 
-        return response;
+      if (!(projectUsers.Any() || _accessValidator.IsAdmin()))
+      {
+        return new FindResultResponse<TaskInfo>();
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:123]: just new()

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 6]
```diff
@@ -152,21 +129,23 @@ public FindResultResponse<TaskInfo> Execute(FindTasksFilter filter)
       List<Guid> users = dbTasks.Where(x => x.AssignedTo.HasValue).Select(x => x.AssignedTo.Value).ToList();
       users.AddRange(dbTasks.Select(x => x.CreatedBy).ToList());
 
-      IGetUsersDataResponse usersData = GetUsersData(users, response.Errors);
+      List<UserData> usersData = await GetUsersDatas(users, errors);
 
       List<TaskInfo> tasks = new();
-      foreach (DbTask dbTask in dbTasks)
+      foreach (var dbTask in dbTasks)
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:135]: var

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 7]
```diff
@@ -40,15 +44,30 @@ public class GetTaskCommand : IGetTaskCommand
     private readonly ILogger<GetTaskCommand> _logger;
     private readonly IRequestClient<IGetCompanyEmployeesRequest> _rcGetCompanyEmployee;
     private readonly IRequestClient<IGetUsersDataRequest> _usersDataRequestClient;
+    private readonly IConnectionMultiplexer _cache;
     private readonly IRequestClient<IGetImagesRequest> _rcImages;
     private readonly IImageInfoMapper _imageMapper;
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:49]: remove _rcImages and _imageMapper

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 8]
```diff
@@ -40,15 +44,30 @@ public class GetTaskCommand : IGetTaskCommand
...
+
+    private async Task<DepartmentData> GetDepartmentThroughBroker(Guid authorId, List<string> errors)
+    {
+      string errorMessage = "Cannot get department. Please try again later.";
+
       try
       {
-        Response<IOperationResult<IGetCompanyEmployeesResponse>> response = _rcGetCompanyEmployee.GetResponse<IOperationResult<IGetCompanyEmployeesResponse>>(
-          IGetCompanyEmployeesRequest.CreateObj(new() { authorId }, includeDepartments: true)).Result;
+        var response = await _rcGetCompanyEmployee.GetResponse<IOperationResult<IGetCompanyEmployeesResponse>>(
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:69]: var

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 9]
```diff
@@ -145,29 +146,31 @@ private List<ImageInfo> GetTaskImages(List<Guid> imageIds, List<string> errors)
...
-      if (task.AssignedTo.HasValue)
-      {
-        userIds.Add(task.AssignedTo.Value);
-      }
+      Guid requestUserId = _httpContext.GetUserId();
 
-      if (task.ParentTask != null)
+      DepartmentData department = await GetDepartment(requestUserId, errors);
+
+      if (department != null)
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:160]: if -> if == if && if

Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:160]: ((if department != null && department.DirectorUserId == requestUserId) || _accessValidator.IsAdmin(requestUserId)) -> return (true, department)

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 10]
```diff
@@ -192,46 +194,57 @@ private List<Guid> GetUsersIds(DbTask task, Guid? parentTaskAssignedTo)
...
-      _rcImages = rcImages;
-      _imageMapper = imageMapper;
+      _cache = cache;
     }
 
-    public OperationResultResponse<TaskResponse> Execute(Guid taskId, bool isFullModel = true)
+    public async Task<OperationResultResponse<TaskResponse>> Execute(Guid taskId, bool isFullModel = true)
     {
-      OperationResultResponse<TaskResponse> response = new OperationResultResponse<TaskResponse>();
+      var errors = new List<string>();
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:202]: var

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 11]
```diff
@@ -192,46 +194,57 @@ private List<Guid> GetUsersIds(DbTask task, Guid? parentTaskAssignedTo)
...
-        response.Errors.Add($"Task with Id {taskId} was not found.");
+      if (!hasRights)
+      {
+        _httpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
 
-        return response;
+        return new OperationResultResponse<TaskResponse>
+        {
+          Status = OperationResultStatusType.Failed,
+          Errors = new() { "Not enough rights." }
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:215]: errors already instantiated -> use add. i think it faster then create new object

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 12]
```diff
@@ -192,46 +194,57 @@ private List<Guid> GetUsersIds(DbTask task, Guid? parentTaskAssignedTo)
...
+        };
       }
 
-      if (!Authorization(task.ProjectId, out DepartmentData department))
+      List<Guid> userIds = new()
       {
-        _httpContext.Response.StatusCode = (int)HttpStatusCode.Forbidden;
+        task.CreatedBy,
+      };
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:222]: write it in 1 string

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs [hunk 13]
```diff
@@ -192,46 +194,57 @@ private List<Guid> GetUsersIds(DbTask task, Guid? parentTaskAssignedTo)
...
+      List<UserData> usersDataResponse = GetUsersInfo(userIds, errors);
 
-      List<ImageInfo> imagesInfo = GetTaskImages(task.Images.Select(x => x.ImageId).ToList(), response.Errors);
+      List<ImageInfo> imagesInfo = GetTaskImages(task.Images.Select(x => x.ImageId).ToList(), errors);
 
       List<TaskInfo> subtasksInfo = new();
       if (task.Subtasks != null)
       {
-        foreach (DbTask dbSubtask in task.Subtasks)
+        foreach (var dbSubtask in task.Subtasks)
```
Comment (Glunky) [inline src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:247]: var

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 7 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:49]: ??? why? we use it here

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs (hunk 1 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 2 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 3 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs]: mistake in merge, return code from prev commit

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 4 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 7 above)
Comment (Glunky) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:49]: sorry

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 5 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs]: mistake in merge, return code from prev commit

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 6 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 8 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 9 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: reworked

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 10 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 11 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 12 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 13 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 7 above)
Comment (Glunky) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:49]: sorry. but in not instanticated -  you removed
_rcImages = rcImages;
_imageMapper = imageMapper;
see warnings.

**Review by Glunky — APPROVED** [review]

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs [hunk 14]
```diff
@@ -49,9 +65,9 @@ private List<DepartmentData> GetDepartments(List<DbProject> dbProjects, List<str
 
       try
       {
-        Response<IOperationResult<IGetDepartmentsResponse>> response = _rcGetDepartments
-          .GetResponse<IOperationResult<IGetDepartmentsResponse>>(
-            IGetDepartmentsRequest.CreateObj(departmentIds)).Result;
+        Response<IOperationResult<IGetDepartmentsResponse>> response = await _rcGetDepartments
+            .GetResponse<IOperationResult<IGetDepartmentsResponse>>(
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs:69]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 15]
```diff
@@ -21,289 +29,340 @@
...
+
+      return await GetDepartmentThroughBroker(departmentId, errors);
+    }
+
+    private async Task<DepartmentData> GetDepartmentThroughBroker(Guid departmentId, List<string> errors)
+    {
+      try
+      {
+        Response<IOperationResult<IGetDepartmentsResponse>> departmentResponse =
+            await _rcGetDepartment.GetResponse<IOperationResult<IGetDepartmentsResponse>>(
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:70]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 16]
```diff
@@ -21,289 +29,340 @@
...
-
-            errors.Add($"Can not get department info for DepartmentId '{departmentId}'. Please try again later.");
-
-            return null;
+          return departmentResponse.Message.Body.Departments.FirstOrDefault();
         }
 
-        private List<ImageInfo> GetUserAvatars(List<Guid> imageIds, List<string> errors)
+        _logger.LogWarning(
+            $"Can not get department. Reason:{Environment.NewLine}{string.Join('\n', departmentResponse.Message.Errors)}.");
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:79]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 17]
```diff
@@ -21,289 +29,340 @@
...
+      }
+
+      errors.Add($"Can not get department info for DepartmentId '{departmentId}'. Please try again later.");
+
+      return null;
+    }
+
+    private async Task<List<ImageInfo>> GetUserAvatars(List<Guid> imageIds, List<string> errors)
+    {
+      if (imageIds == null || imageIds.Count == 0)
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:93]: imageIds.Count => !ImageIds.Any()

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 18]
```diff
@@ -21,289 +29,340 @@
...
+      {
+        return null;
+      }
+
+      string logMessage = "Errors while getting images with ids: {Ids}. Errors: {Errors}";
+
+      try
+      {
+        Response<IOperationResult<IGetImagesResponse>> response = await _rcImages.GetResponse<IOperationResult<IGetImagesResponse>>(
+            IGetImagesRequest.CreateObj(imageIds, ImageSource.User));
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:103]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 19]
```diff
@@ -21,289 +29,340 @@
...
 
-            List<Guid> usersIds = projectUsers.Select(x => x.UserId).ToList();
+      return null;
+    }
 
-            return GetProjectUsers(usersIds, projectUsers, ref errors);
-        }
+    private async Task<List<ImageInfo>> GetProjectImages(List<Guid> imageIds, List<string> errors)
+    {
+      if (imageIds == null || imageIds.Count == 0)
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:129]: same

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 20]
```diff
@@ -21,289 +29,340 @@
...
+      {
+        return null;
+      }
+
+      string logMessage = "Errors while getting images with ids: {Ids}. Errors: {Errors}";
+
+      try
+      {
+        Response<IOperationResult<IGetImagesResponse>> response = await _rcImages.GetResponse<IOperationResult<IGetImagesResponse>>(
+           IGetImagesRequest.CreateObj(imageIds, ImageSource.Project));
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:139]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 21]
```diff
@@ -21,289 +29,340 @@
...
-            _rcGetDepartment = rcGetDepartments;
-            _usersDataRequestClient = usersDataRequestClient;
-            _rcGetCompanyEmployees = rcGetCompanyEmployees;
-            _rcImages = rcImages;
-            _httpContextAccessor = httpContextAccessor;
+          return usersDataResponse.Message.Body.UsersData;
         }
+        
+        _logger.LogWarning(
+            $"Can not get users. Reason:{Environment.NewLine}{string.Join('\n', usersDataResponse.Message.Errors)}.");
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:196]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 22]
```diff
@@ -21,289 +29,340 @@
...
+            IGetCompanyEmployeesRequest.CreateObj(usersIds, includeDepartments: includeDepartments, includePositions: includePositions));
+
+        if (response.Message.IsSuccess)
+        {
+          return response.Message.Body;
+        }
+        else
+        {
+          _logger.LogWarning("Errors while getting users departments and positions for users {UserIds}. Reason: {Errors}",
+              string.Join(", ", usersIds),
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:274]: tab

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 23]
```diff
@@ -21,289 +29,340 @@
...
+              string.Join('\n', response.Message.Errors));
         }
+      }
+      catch (Exception exc)
+      {
+        _logger.LogError(exc,
+            "Can not get user's departments and positions for users {UserIds}. Please try again later.",
```
Comment (Cheephot) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:281]: tab

**Review by Cheephot — APPROVED** [review]

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 23 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 22 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 21 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 20 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 19 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 18 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 17 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 16 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 15 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs (hunk 14 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs]: fixed

File: src/ProjectService.Business/Commands/Task/GetTaskCommand.cs (hunk 7 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/GetTaskCommand.cs:49]: fixed

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs [hunk 24]
```diff
@@ -99,13 +117,12 @@ public FindResultResponse<ProjectInfo> Execute(FindProjectsFilter filter)
         };
       }
 
-      FindResultResponse<ProjectInfo> response = new();
-
-      List<DbProject> dbProjects = _repository.Find(filter, out int totalCount);
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs:104]: Why do you rename it?

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 25]
```diff
@@ -21,289 +29,340 @@
...
-            errors.Add($"Can not get users info for UserIds {string.Join('\n', usersIds)}. Please try again later.");
-
-            return null;
+          return response.Message.Body.ImagesData.Select(_imageMapper.Map).ToList();
         }
-
-        private IGetCompanyEmployeesResponse GetCompanyEmployees(
-            List<Guid> userIds,
-            List<string> errors)
+        else
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:145]: what do you use else?

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 26]
```diff
@@ -21,289 +29,340 @@
...
+      }
+
+      List<UserData> usersDatas = await GetUsersDatas(dbProject.Users, response.Errors);
+      List<ProjectUserInfo> usersInfo = null;
+      List<Guid> usersIds = dbProject.Users.Select(u => u.UserId).Distinct().ToList();
+
+      if (usersDatas != null && usersDatas.Any())
+      {
+        (List<DepartmentData> departments, List<PositionData> positions) = await GetCompanyEmployee(usersIds, response.Errors);
+        List<ImageInfo> imagesInfos = await GetUserAvatars(usersDatas.Where(u => u.ImageId.HasValue).Select(u => u.ImageId.Value).ToList(), response.Errors);
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:338]: maybe you add ?operation:  ....(usersDatas?.Where(u =>....

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 27]
```diff
@@ -21,289 +29,340 @@
...
+      DbProject dbProject = _repository.Get(filter);
+
+      if (dbProject.DepartmentId.HasValue)
+      {
+        department = await GetDepartment(dbProject.DepartmentId.Value, response.Errors);
+      }
+
+      List<UserData> usersDatas = await GetUsersDatas(dbProject.Users, response.Errors);
+      List<ProjectUserInfo> usersInfo = null;
+      List<Guid> usersIds = dbProject.Users.Select(u => u.UserId).Distinct().ToList();
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:333]: List<Guid> usersIds = dbProject.Users?.Sel.....

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 28]
```diff
@@ -21,289 +29,340 @@
...
+        //rework
+        List<DbProjectUser> projectUsersForCount = _userRepository.Find(usersIds);
+
+        usersInfo = dbProject.Users
+          .Select(pu =>
+          {
+            UserData mappedUser = usersDatas.FirstOrDefault(x => x.Id == pu.UserId);
+
+            return _projectUserInfoMapper.Map(
+              mappedUser,
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:349]: usersDatas?.FirstOrDefault(x => x.Id == pu.UserId)

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 29]
```diff
@@ -21,289 +29,340 @@
...
+              imagesInfos?.FirstOrDefault(i => i.Id == mappedUser.ImageId),
+              positions?.FirstOrDefault(p => p.UsersIds.Any(id => id == pu.UserId)),
+              departments?.FirstOrDefault(d => d.UsersIds.Any(id => id == pu.UserId)),
+              pu,
+              projectUsersForCount.Where(u => u.UserId == pu.UserId).Count());
+          })
+          .ToList();
+      }
+
+      List<ProjectFileInfo> filesInfo = dbProject.Files.Select(_projectFileInfoMapper.Map).ToList();
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:359]: Files?.

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs [hunk 30]
```diff
@@ -21,289 +29,340 @@
...
+      List<ImageInfo> imagesinfo = await GetProjectImages(dbProject.Images.Select(x => x.ImageId).ToList(), response.Errors);
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:360]: Images?.

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs [hunk 31]
```diff
@@ -126,7 +149,7 @@ public FindResultResponse<TaskInfo> Execute(FindTasksFilter filter)
       if (!_accessValidator.IsAdmin()
```
Comment (MariyaSin) [inline src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:149]: you can remove isAdmin check, you have right chesk

File: src/ProjectService.Models.Dto/Responses/FindResponse.cs [hunk 32]
```diff
@@ -2,10 +2,10 @@
...
 {
-    public class FindResponse<T>
-    {
-        public int TotalCount { get; set; }
-        public IEnumerable<T> Body { get; set; } = new List<T>();
-        public IEnumerable<string> Errors { get; set; } = new List<string>();
-    }
-}
\ No newline at end of file
+  public class FindResponse<T>
```
Comment (MariyaSin) [inline src/ProjectService.Models.Dto/Responses/FindResponse.cs:5]: we use FindResultResponse from kernel

**Review by MariyaSin — APPROVED** [review]

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 26 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:338]: if check that usersData is not null

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 25 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:145]: removed

File: src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs (hunk 24 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/FindProjectsCommand.cs:104]: renamed back

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 27 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:333]: is not null

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 28 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:349]: used for image id

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 29 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:359]: is not null

File: src/ProjectService.Business/Commands/Project/GetProjectCommand.cs (hunk 30 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Project/GetProjectCommand.cs:360]: is not null

File: src/ProjectService.Business/Commands/Task/FindTasksCommand.cs (hunk 31 above)
Comment (BelehovEgor) [reply src/ProjectService.Business/Commands/Task/FindTasksCommand.cs:149]: fixed

File: src/ProjectService.Models.Dto/Responses/FindResponse.cs (hunk 32 above)
Comment (BelehovEgor) [reply src/ProjectService.Models.Dto/Responses/FindResponse.cs:5]: reworked

**Pull request merged**